File: C++/TwoPointers.cpp
#include <iostream>
#include <vector>
#include <map>
#include <cmath>
#include <algorithm>
#include <deque>
#include <queue>
#include <stack>
#include <set>
#include <string>
#include <string.h>
#include <array>
#include <unordered_map>
#include <unordered_set>
#include <iomanip>
using namespace std;
typedef long long ll;
typedef long double ld;
#define pb push_back

    // O(n)
ll calculate(int n, ll k, ll arr[]) {
    ll ans=0;
    
        // one = left one, two = right one
    queue<ll> one, onemin, onemax, two, twomin, twomax;
    
    for (int i=0; i<n; i++) {
        two.push(arr[i]);
        if (twomin.size() > 0) twomin.push(min(arr[i], twomin.front()));
        else twomin.push(arr[i]);
        if (twomax.size() > 0) twomax.push(max(arr[i], twomax.front()));
        else twomax.push(arr[i]);
        
        if (one.empty()) {
            // move everything over to one
            while (!two.empty()) {
                one.push(two.front());
                two.pop();
                twomin.pop(); twomax.pop();
                if (onemin.size() > 0) onemin.push(min(one.front(), onemin.front()));
                else onemin.push(one.front());
                if (onemax.size() > 0) onemax.push(max(one.front(), onemax.front()));
                else onemax.push(one.front());
            }	
        }
        
        ll mn = onemin.front();
        if (twomin.size() > 0) mn = min(mn, twomin.front());
        ll mx = onemax.front();
        if (twomax.size() > 0) mx = max(mx, twomax.front());
        
        while (mx - mn > k) {
            one.pop();
            onemin.pop(); onemax.pop();
            
            if (one.empty()) {
                // move everything over to one
                while (!two.empty()) {
                    one.push(two.front());
                    two.pop();
                    twomin.pop(); twomax.pop();
                    if (onemin.size() > 0) onemin.push(min(one.front(), onemin.front()));
                    else onemin.push(one.front());
                    if (onemax.size() > 0) onemax.push(max(one.front(), onemax.front()));
                    else onemax.push(one.front());
                }
            }
            
            mn = onemin.front();
            if (twomin.size() > 0) mn = min(mn, twomin.front());
            mx = onemax.front();
            if (twomax.size() > 0) mx = max(mx, twomax.front());
        }
        
        ans = max(ans, (ll)(one.size() + two.size()));
    }
    return ans;
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);

    
}

File: Java/TwoPointers.java

import java.util.*;
import java.io.*;

public class TwoPointers {

	public static void main(String[] args) {

	}
	
	// O(n)
	public static long calculate(int n, long k, long[] arr) {
		long ans = 0;
		
			// one = left one, two = right one
		ArrayDeque<Long> one = new ArrayDeque<>(), oneMin = new ArrayDeque<>(), 
				oneMax = new ArrayDeque<>(), two = new ArrayDeque<>(), 
				twoMin = new ArrayDeque<>(), twoMax = new ArrayDeque<>();		
		
		for (int i = 0; i < n; i++) {
			two.push(arr[i]);
			if (twoMin.size() > 0) twoMin.push(Math.min(arr[i], twoMin.peek()));
			else twoMin.push(arr[i]);
			if (twoMax.size() > 0) twoMax.push(Math.max(arr[i], twoMax.peek()));
			else twoMax.push(arr[i]);
			
			if (one.isEmpty()) {
				moveFromTwoToOne(one, oneMin, oneMax, two, twoMin, twoMax);
			}
			
			long min = oneMin.peek();
			if (twoMin.size() > 0) min = Math.min(min, twoMin.peek());
			long max = oneMax.peek();
			if (twoMax.size() > 0) max = Math.max(max, twoMax.peek());
			
			while (max - min > k) {
				one.pop();
				oneMin.pop(); oneMax.pop();
				
				if (one.isEmpty()) {
					moveFromTwoToOne(one, oneMin, oneMax, two, twoMin, twoMax);
				}
				
				min = oneMin.peek();
				if (twoMin.size() > 0) min = Math.min(min, twoMin.peek());
				max = oneMax.peek();
				if (twoMax.size() > 0) max = Math.max(max, twoMax.peek());
			}
			
			ans = Math.max(ans, one.size() + two.size());
		}
		return ans;
	}

	public static void moveFromTwoToOne(ArrayDeque<Long> one, ArrayDeque<Long> oneMin, ArrayDeque<Long> oneMax,
											ArrayDeque<Long> two, ArrayDeque<Long> twoMin, ArrayDeque<Long> twoMax) {
		// move everything over from two to one
		while (!two.isEmpty()) {
			one.push(two.pop());
			twoMin.pop(); twoMax.pop();
			if (oneMin.size() > 0) oneMin.push(Math.min(one.peek(), oneMin.peek()));
			else oneMin.push(one.peek());
			if (oneMax.size() > 0) oneMax.push(Math.max(one.peek(), oneMax.peek()));
			else oneMax.push(one.peek());
		}	
	}
}